File: apps/backend/src/workers/evaluation.worker.ts
import { Injectable, Logger } from '@nestjs/common';
import { Cron, CronExpression } from '@nestjs/schedule';
import { EvaluationJobService } from '../modules/evaluation-job/evaluation-job.service';
import { EvaluationService } from '../modules/evaluation/evaluation.service';

/**
 * EvaluationWorker — cron-based async evaluation processor.
 *
 * Design decisions:
 * - Polls every 10s for PENDING jobs that are ready (respects nextRetryAt backoff)
 * - isRunning guard prevents overlapping executions on the same instance
 * - Handles idempotency: if the session was already evaluated (ConflictException),
 *   the job is marked COMPLETED rather than FAILED — crash-safe (spec §5)
 * - Separate zombie recovery cron resets PROCESSING jobs stuck >10min (spec §6)
 * - Processes up to N_CONCURRENT jobs per tick to drain backlogs faster without
 *   overwhelming the OpenAI rate limits
 */
@Injectable()
export class EvaluationWorker {
    private readonly logger = new Logger(EvaluationWorker.name);
    private isRunning = false;

    // Max jobs to process per tick — balances throughput vs rate limit safety
    private readonly BATCH_SIZE = 3;

    constructor(
        private readonly evaluationJobService: EvaluationJobService,
        private readonly evaluationService: EvaluationService,
    ) { }

    // ── Main processing loop — every 10s ─────────────────────────────────────
    @Cron(CronExpression.EVERY_10_SECONDS)
    async processNextJobs() {
        if (this.isRunning) {
            this.logger.debug('Worker tick skipped — previous run still active');
            return;
        }
        this.isRunning = true;

        try {
            let processed = 0;

            // Process up to BATCH_SIZE jobs per tick
            while (processed < this.BATCH_SIZE) {
                const job = await this.evaluationJobService.claimNextPendingJob();
                if (!job) break; // queue empty or all remaining jobs are in backoff window

                processed++;
                const startTime = Date.now();
                this.logger.log(
                    `Processing EvaluationJob ${job.id} | Session: ${job.sessionId} | ` +
                    `Attempt: ${job.attempts + 1}`,
                );

                try {
                    await this.evaluationService.analyzeSessionInternal(job.sessionId);
                    const durationMs = Date.now() - startTime;
                    await this.evaluationJobService.markCompleted(job.id);
                    this.logger.log(
                        `EvaluationJob ${job.id} COMPLETED in ${durationMs}ms`,
                    );
                } catch (err: unknown) {
                    await this.handleJobError(job.id, job.sessionId, err);
                }
            }

            if (processed > 0) {
                this.logger.debug(`Worker tick processed ${processed} job(s)`);
            }
        } finally {
            this.isRunning = false;
        }
    }

    // ── Zombie recovery cron — every 5 minutes (spec §6) ─────────────────────
    // Finds PROCESSING jobs stuck >10min (worker crashed mid-execution) and
    // resets them to PENDING so they'll be retried with a 30s delay.
    @Cron('0 */5 * * * *')  // every 5 minutes
    async recoverZombieJobs() {
        const count = await this.evaluationJobService.recoverZombieJobs();
        if (count > 0) {
            this.logger.warn(`Zombie recovery: reset ${count} stuck job(s) to PENDING`);
        }
    }

    /**
     * Handle job error with idempotency protection (spec §5).
     *
     * ConflictException means the EvaluationReport was already saved
     * (worker crashed after saving report but before markCompleted).
     * In this case, marking the job COMPLETED is correct, not FAILED.
     *
     * All other errors → markFailed (which applies exponential backoff).
     */
    private async handleJobError(
        jobId: string,
        sessionId: string,
        err: unknown,
    ): Promise<void> {
        const message = err instanceof Error ? err.message : String(err);

        // ConflictException = report already exists = evaluation was actually successful
        // This is the idempotency recovery path (spec §5)
        const isAlreadyEvaluated =
            message.includes('Evaluation already exists') ||
            message.includes('ConflictException') ||
            message.includes('Unique constraint');

        if (isAlreadyEvaluated) {
            this.logger.warn(
                `EvaluationJob ${jobId} (session ${sessionId}): ` +
                `report already exists — marking COMPLETED (idempotent recovery)`,
            );
            await this.evaluationJobService.markCompleted(jobId);
            return;
        }

        // Real failure — apply exponential backoff retry
        this.logger.error(
            `EvaluationJob ${jobId} (session ${sessionId}) failed: ${message}`,
        );
        await this.evaluationJobService.markFailed(jobId, message);
    }
}

File: apps/api/app/workers/evaluation_worker.py
"""
EvaluationWorker — background job that processes pending evaluation jobs.

Schedule (via APScheduler in scheduler.py):
  - run_evaluation_tick()  every 10 seconds — claims and processes one PENDING job
  - recover_zombie_jobs()  every 5 minutes  — resets stuck PROCESSING jobs to PENDING

Design decisions:
  - Each tick claims exactly one job (SELECT ... FOR UPDATE SKIP LOCKED)
  - Idempotency: if the session was already analyzed (ConflictException from service),
    the worker marks the job COMPLETED rather than FAILED
  - Worker uses its own DB session (not the request-scoped session)
  - All failures are caught; the job is marked FAILED with backoff
  - Zombie recovery is separate from the main tick to keep error surfaces isolated

Matches NestJS: apps/backend/src/modules/evaluation-job/evaluation-job.service.ts (worker logic)
"""
import logging
import uuid

from app.db.session import SessionLocal
from app.modules.evaluation import repository as eval_repo
from app.modules.evaluation import service as eval_service

logger = logging.getLogger(__name__)


async def run_evaluation_tick() -> None:
    """
    Called every 10 seconds by the APScheduler.
    Claims one PENDING job, runs evaluation, marks it COMPLETED or FAILED.
    """
    async with SessionLocal() as db:
        try:
            job = await eval_repo.claim_next_pending_job(db)
            if job is None:
                return  # nothing to process right now

            session_id: uuid.UUID = job.session_id
            job_id: uuid.UUID = job.id
            logger.info("EvaluationWorker: processing job %s for session %s", job_id, session_id)

            try:
                await eval_service.analyze_session_internal(db, session_id)
                await eval_repo.mark_job_completed(db, job_id)
                await db.commit()
                logger.info("EvaluationWorker: job %s COMPLETED", job_id)

            except Exception as exc:
                # Idempotency: ConflictException means evaluation already exists
                # (worker crashed after report was saved but before markCompleted).
                # Mark COMPLETED rather than retrying.
                from app.core.exceptions import ConflictException
                if isinstance(exc, ConflictException):
                    logger.warning(
                        "EvaluationWorker: job %s — evaluation already exists, marking COMPLETED",
                        job_id,
                    )
                    await db.rollback()
                    async with SessionLocal() as fix_db:
                        await eval_repo.mark_job_completed(fix_db, job_id)
                        await fix_db.commit()
                    return

                # All other errors → mark failed with backoff
                error_msg = str(exc)
                logger.error(
                    "EvaluationWorker: job %s FAILED — %s", job_id, error_msg
                )
                await db.rollback()
                async with SessionLocal() as fail_db:
                    await eval_repo.mark_job_failed(fail_db, job_id, error_msg)
                    await fail_db.commit()

        except Exception as exc:
            # Errors claiming the job itself (rare DB issues)
            logger.error("EvaluationWorker: unexpected error in tick — %s", exc)
            try:
                await db.rollback()
            except Exception:
                pass


async def run_zombie_recovery() -> None:
    """
    Called every 5 minutes by the APScheduler.
    Resets PROCESSING jobs that have been stuck for >10 minutes.
    """
    async with SessionLocal() as db:
        try:
            count = await eval_repo.recover_zombie_jobs(db)
            await db.commit()
            if count > 0:
                logger.warning("Zombie recovery: reset %d job(s)", count)
        except Exception as exc:
            logger.error("Zombie recovery failed: %s", exc)
            await db.rollback()
